ada/language/ast.py

```diff
@@ -6159,24 +6169,26 @@ class BaseAggregate(Expr):
         Self, Predicate(BaseTypeDecl.is_array_or_rec, Self.type_var)
     ))
 
-    @langkit_property(return_type=T.BaseAggregate, ignore_warn_on_node=True)
-    def root_direct_parent_aggregate():
-        return Self.direct_parent_aggregate.then(
-            lambda p: p.root_direct_parent_aggregate,
-            default_val=Self
-        )
-
-    @langkit_property(return_type=T.Int)
-    def sub_aggregate_rank(r=(Int, 0)):
-        return Self.direct_parent_aggregate.then(
-            lambda p: p.sub_aggregate_rank(r + 1),
-            default_val=r
+    @langkit_property(return_type=MultidimAggregateInfo, dynamic_vars=[origin])
+    def multidim_root_aggregate(r=(Int, 0)):
+        """
+        Return the root parent aggregate if Self is part of a multidimensional
+        array aggregate (either the root or a sub-aggregate).
+        """
+        return Entity.type_val.cast(BaseTypeDecl).then(
```

Comment:
Can you explain this logic? What kind of value can "type_val" take for nested aggregates?

Reply:
nested aggregates of a multidimensional array have no types, so type_val is empty

Reply:
Added some comments !